src/sipo.rs

```diff
@@ -1,36 +1,40 @@
 //! Serial-in parallel-out shift register
 
 use core::cell::RefCell;
-use hal::digital::OutputPin;
-use core::mem;
-use core::ptr;
+use core::mem::{self, MaybeUninit};
+
+use crate::hal::digital::v2::OutputPin;
 
 trait ShiftRegisterInternal {
-    fn update(&self, index: usize, command: bool);
+    fn update(&self, index: usize, command: bool) -> Result<(), ()>;
 }
 
 /// Output pin of the shift register
 pub struct ShiftRegisterPin<'a>
 {
-    shift_register: &'a ShiftRegisterInternal,
+    shift_register: &'a dyn ShiftRegisterInternal,
     index: usize,
 }
 
 impl<'a> ShiftRegisterPin<'a>
 {
-    fn new(shift_register: &'a ShiftRegisterInternal, index: usize) -> Self {
+    fn new(shift_register: &'a dyn ShiftRegisterInternal, index: usize) -> Self {
         ShiftRegisterPin { shift_register, index }
     }
 }
 
-impl<'a> OutputPin for ShiftRegisterPin<'a>
+impl OutputPin for ShiftRegisterPin<'_>
 {
-    fn set_low(&mut self) {
-        self.shift_register.update(self.index, false);
+    type Error = ();
```

Comment:
Is it feasible to use the error type from whichever OutputPins the shift register uses internally here, so we can raise up actual errors instead of mapping them to `()`?

Reply:
Well, the problem at the moment is, that `ShiftRegisterPin` would have to be generic over the three Pins instead of having a `&dyn ShiftRegisterInternal` in order to design an error like the following

```enum ShiftRegisterError<E1, E2, E3> {
    ClockPin(E1),
    LatchPin(E2),
    DataPin(E3),
}

impl<Pin1, Pin2, Pin3> OutpitPin for ShiftRegisterPin<Pin1, Pin2, Pin3> {
    type Error = ShiftRegisterError<
        Pin1::Error,
        Pin2::Error,
        Pin3::Error,
    >

    // ignoring trait functions here
}
```
We could still return a string error stating which Pin failed but we decided that that's kind of pointless on embedded and will anyways only be detected during debugging where it will be obvious which pin failed. 
If you have another idea I would be delighted to hear it. I'm not a 100 percent sure, why the `ShiftRegisterInternal` trait exists, but I think it has something to do with the ability to split up the `ShiftRegister` into `ShiftRegisterPins`.

Reply:
Shift registers usually do not return any type of acknowledging or error when we set the output pins, so we'd have no idea when an error occurs. It's kind of pointless to have this Result as return type, but it is required by `v2::OutputPin`